File: morning_exercises/day15.cs
class BSTNode4(int val=0, BSTNode4? left = null, BSTNode4? right = null)
{
    public int val = val;
    public BSTNode4? left = left;
    public BSTNode4? right = right;

    public static int? InOrderSuccessor(BSTNode4? root, int k)
    {
        // if (root==null) return null;

        int? successor = null;

        while (root!=null)
        {
            if (k < root.val)
            {
                successor=root.val;
                root=root.left;
            }
            else root=root.right;
        }
        return successor;
    }
}

File: morning_exercises/day15.ts
class BSTNode4 {
    constructor(
        public val: number = 0,
        public left: BSTNode4 | null = null,
        public right: BSTNode4 | null = null
    ) {}

    static inOrderSuccessor(root: BSTNode4 | null, k: number): number | null {
        if (root==null) return null;

        let successor: number | null = null;

        while (root) {
            if (k<root.val) {
                successor = root.val;
                root=root.left;
            }
            else root=root.right;
        }
        return successor;
    }
}

{
    const root = new BSTNode4(6);
    root.left = new BSTNode4(4);
    root.right = new BSTNode4(9);
    root.left.left = new BSTNode4(2);
    root.left.right = new BSTNode4(5);

    console.log(BSTNode4.inOrderSuccessor(root, 5));
}